src/api/cvc4cpp.h

```diff
@@ -0,0 +1,2357 @@
+     * @param sort the sort of the return value of this function
+     * @return the function
+     */
+    Term declareFun(const std::string& symbol,
+                    const std::vector<Sort>& sorts,
+                    Sort sort) const;
+
+    /**
+     * Declare uninterpreted sort.
+     * SMT-LIB: ( declare-sort <symbol> <numeral> )
+     * @param symbol the name of the sort
+     * @param arity the arity of the sort
+     * @return the sort
+     */
+    Sort declareSort(const std::string& symbol, uint32_t arity) const;
+
+    /**
+     * Define n-ary function.
+     * SMT-LIB: ( define-fun <function_def> )
+     * @param symbol the name of the function
+     * @param bound_vars the parameters to this function
+     * @param sort the sort of the return value of this function
+     * @param term the function body
+     * @return the function
+     */
+    Term defineFun(const std::string& symbol,
+                   const std::vector<Term>& bound_vars,
+                   Sort sort,
+                   Term term) const;
+    /**
+     * Define n-ary function.
+     * SMT-LIB: ( define-fun <function_def> )
+     * Create parameter 'fun' with mkVar().
+     * @param fun the sorted function
+     * @param bound_vars the parameters to this function
+     * @param term the function body
+     * @return the function
+     */
+    Term defineFun(Term fun,
+                   const std::vector<Term>& bound_vars,
+                   Term term) const;
+
+    /**
+     * Define recursive function.
+     * SMT-LIB: ( define-fun-rec <function_def> )
+     * @param symbol the name of the function
+     * @param bound_vars the parameters to this function
+     * @param sort the sort of the return value of this function
+     * @param term the function body
+     * @return the function
+     */
+    Term defineFunRec(const std::string& symbol,
+                      const std::vector<Term>& bound_vars,
+                      Sort sort,
+                      Term term) const;
+
+    /**
+     * Define recursive function.
+     * SMT-LIB: ( define-fun-rec <function_def> )
+     * Create parameter 'fun' with mkVar().
+     * @param fun the sorted function
+     * @param bound_vars the parameters to this function
+     * @param term the function body
+     * @return the function
+     */
+    Term defineFunRec(Term fun,
+                      const std::vector<Term>& bound_vars,
+                      Term term) const;
+
+    /**
+     * Define recursive functions.
+     * SMT-LIB: ( define-funs-rec ( <function_decl>^{n+1} ) ( <term>^{n+1} ) )
+     * Create elements of parameter 'funs' with mkVar().
+     * @param funs the sorted functions
+     * @param bound_vars the list of parameters to the functions
+     * @param term the list of function bodies of the functions
+     * @return the function
+     */
+    void defineFunsRec(const std::vector<Term>& funs,
+                       const std::vector<std::vector<Term>>& bound_vars,
+                  
```

Comment:
I want to open a specific discussion of using PImpl here. This instance has multiple members perhaps this makes sense here even if does not in the other locations.